`ex_act()` will work on basic mobs again (lol) + Unit Test

san7890 opened the pull request:
Regression introduced in #72260
## About The Pull Request

whooo this one's fun

basically ex_act's implementation on basic mobs would call parent and then react to it's value, this is presumably to do the first check about space vine mutations and whatever. the problem is that the `/mob/living` implementation would itself also call parent, and that would always return null because `/atom/proc/ex_act` doesn't have a set return value. So, this simply would _always_ early return, with ex_act presumably *never* working on basic mobs for at least four months now.

I decided to then change up the return values for pretty much all implementations of `ex_act()` since there was no rhyme or reason to returning null/FALSE/TRUE, and documenting why it's like that.

Just to make sure I wasn't breaking anything doing this (at least on base implementations), I wrote a unit test for all of the three major physical types in game (objs, mobs, turfs) because i am a paranoid fuckar. we should be good to go now though.
## Why It's Good For The Game

i noticed this because placing c4's on sargeant araneus wouldn't actually damage it whatsoever. now it actually does the stated 30 damage, but araneus has like 250 health so it doesn't actually matter in the long run. whatever at least it does the damn 30 now.

also adds a unit test for this specific case as well as a range of other cases to ensure this stuff doesn't silently break in this way anymore

## Changelog
:cl:
fix: Basic mobs will now be affected by explosions again.
/:cl:

Proof that the unit test works with the changes made in this PR (the test I wrote failing on a copy of master)
![image](https://user-images.githubusercontent.com/34697715/234141098-14840419-22a8-443b-be84-832c29ba367d.png)

for some reason test_focus shat out but here's proof of it working
![image](https://user-images.githubusercontent.com/34697715/234146814-a60dbbb7-406f-4f0a-bca6-9b241bf48ff1.png)

san7890 committed:
ex_act will now work on basic mobs (lol)

whooo this one's fun

basically ex_act's implementation on basic mobs would call parent and then react to it's value, this is presumably to do the first check about space vine mutations and whatever. the problem is that the /mob/living implementation would itself also call parent, and that would always return null because `/atom/proc/ex_act` doesn't have a set return value. So, this simply would _always_ early return, with ex_act presumably *never* working on basic mobs for a good period of time.

basically I just flipped the logic so that we return TRUE instead of null/FALSE (since those are both falsey values) and mirrored that change to simple_animal logic as well so we can stop doing double the work. actually simple.

i noticed this because placing c4's on sargeant araneus wouldn't actually damage it whatsoever. now it actually does the stated 30 damage, but araneus has like 250 health so it doesn't actually matter in the long run. whatever at least it does the damn 30 now.

san7890 committed:
fixes it for aliens too

san7890 committed:
first implementation

san7890 committed:
cleans up typos

san7890 committed:
don't evaluate alien's overall health

this shouldn't have been here in the first place

san7890 committed:
this is so freaking glongy

san7890 committed:
strips test_focus

san7890 committed:
Merge branch 'exact-unit-testing' into fixes-ex-act

san7890 committed:
no need to adjust the human's health

san7890 committed:
use TEST_ASSERT_EQUAL instead

lol whoopsie

san7890 committed:
changing up the ex_act framework

san7890 committed:
splits out mob logic into it's own proc

san7890 committed:
does turf stuff

san7890 committed:
adds obj tests

san7890 committed:
Update explosion_action.dm

san7890 committed:
some more cleanup and safety

san7890 committed:
this commit sponsored by glup shitto

AAAAAAAAAAAAAAAA

san7890 commented on code/modules/unit_tests/explosion_action.dm:
```diff
@@ -0,0 +1,214 @@
+/datum/unit_test/explosion_action/Run()
+	// We split up this `Run()` into multiple parts based on the over-arching parent type. This is because all of them have different core implementations of `EX_ACT()`, and we want to test all.
+	// All procs also have varying levels of bulkiness to them, and it's valuable to have this level of organization because otherwise it would blend all-together and be an entangled mess.
+	execute_mob_tests()
+	execute_turf_tests()
+	execute_obj_tests()
```
should i split this up into different unit tests in the same file? was unsure

Mothblocks replied:
Meh

Fikou commented on code/game/objects/structures/transit_tubes/transit_tube_pod.dm:
```diff
@@ -54,8 +54,11 @@
 	. = ..()
-	if(!QDELETED(src))
-		empty_pod()
+	if(QDELETED(src))
+		return FALSE
```
how is this resisting the explosion? it was deleted due to the explosion, why wouldnt it return true?

Fikou replied:
> how is this resisting the explosion? this checks if it was deleted due to the explosion from parent, why wouldnt it return true?

san7890 commented on code/game/objects/structures/transit_tubes/transit_tube_pod.dm:
```diff
@@ -54,8 +54,11 @@
 	. = ..()
-	if(!QDELETED(src))
-		empty_pod()
+	if(QDELETED(src))
+		return FALSE
```
i think it's kinda weird because we aren't doing any actual work as far as deleting ourselves is concerned but sure
```suggestion
	if(QDELETED(src))
		return TRUE
```

san7890 committed:
Update code/game/objects/structures/transit_tubes/transit_tube_pod.dm

Fikou commented on code/game/objects/obj_defense.dm:
```diff
@@ -5,14 +5,14 @@
 	. = ..() //contents explosion
 	if(QDELETED(src))
-		return
+		return FALSE
```
Same here, and other places probably.

san7890 committed:
this should be all of them

Mothblocks commented on code/modules/unit_tests/explosion_action.dm:
```diff
@@ -0,0 +1,214 @@
+/// that may be done on the subtype-to-subtype basis. Any time we use an explicit subtype is to test that framework, so if you update that for some reason, you should also update this test.
+/// Like, if you balance aliens to take more ear damage and this test fails, just update the test to reflect that. That's it.
+/datum/unit_test/explosion_action/proc/execute_mob_tests()
+
```
```suggestion
```

Mothblocks commented on code/modules/unit_tests/explosion_action.dm:
```diff
@@ -0,0 +1,214 @@
+	return returnable_dog
+
+/// Proc to lessen the amount of copypasta we do for the alien tests, simply sets the rolling vars we have.
+/datum/unit_test/explosion_action/proc/get_alien_damages(mob/living/carbon/alien/subject)
```
Nit: `read_alien_damages` or something might be nicer because `get_` implies a return value

Mothblocks commented on code/modules/unit_tests/explosion_action.dm:
```diff
@@ -0,0 +1,214 @@
+#undef CLOSED_FLOOR_TYPE
+
+/// From the base implementation of `/atom`, [var/max_integrity].
+#define MAX_INTEGRITY_VALUE 500
```
Why not `initial(atom_thing.max_integrity)`?

If we were on 515 right now it'd be `/atom::max_integrity`

san7890 replied:
i wasn't sure how picky we should be since atom_integrity is considered a private var? i've got a better idea anyhow

Mothblocks commented on code/game/atoms.dm:
```diff
@@ -915,9 +915,16 @@
+ * Returns TRUE by default, and behavior should be implemented on children procs on a per-atom basis. Should only return FALSE if we resist the explosion for any reason.
+ * We assume that the default is TRUE because all atoms should be considered destructible in some manner unless they explicitly opt out (in our current framework).
+ * However, the return value itself doesn't have any external consumers, it's only so children procs can listen to the value from their parent procs (due to the nature of the [EX_ACT] macro).
+ * Thus, the return value only matters on overrides of this proc, and the only thing that truly matters is the code that is executed (applying damage, calling damage procs, etc.)
```
We should stack_trace if the return value isn't true/false from the caller, like we do for balloon alerts.

san7890 replied:
i looked for a bit in balloon alert code and couldn't find what you were referring to, what line/file has the stack_trace?

san7890 committed:
get -> read

also gets rid of errant newline

san7890 committed:
cached_max_integrity implementation

we don't need initial() here let's save the time to look it up. also much more hardy

san7890 committed:
minor spelling mistake
